File: test/black-box_test/python/latch.py
'''
 This library offers an API to use Latch in a python environment.
 Copyright (C) 2013 Eleven Paths

 This library is free software; you can redistribute it and/or
 modify it under the terms of the GNU Lesser General Public
 License as published by the Free Software Foundation; either
 version 2.1 of the License, or (at your option) any later version.
 
 This library is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 Lesser General Public License for more details.
 
 You should have received a copy of the GNU Lesser General Public
 License along with this library; if not, write to the Free Software
 Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
'''


import json
import logging

class Error(object):

    def __init__(self, json_data):
        '''
        Constructor
        '''
        
        self.code = json_data['code']
        self.message = json_data['message'] 
        
        
    def get_code(self):
        return self.code
    
    def get_message(self):
        return self.message
    
    def to_json(self):
        return {"code" : self.code, "message" : self.message}
    
    def __repr__(self):
        return json.dumps(self.to_json())
    
    def __str__(self):
        return self.__repr__()
  

   
class LatchResponse(object):
    '''
    This class models a response from any of the endpoints in the Latch API.
    It consists of a "data" and an "error" elements. Although normally only one of them will be
    present, they are not mutually exclusive, since errors can be non fatal, and therefore a response
    could have valid information in the data field and at the same time inform of an error.
    '''

    def __init__(self, json_string):
        '''
        @param $json a json string received from one of the methods of the Latch API
        '''
        json_object = json.loads(json_string)
        if "data" in json_object:
            self.data = json_object["data"]
        else:
            self.data = ""
        
        if "error" in json_object:
            self.error = Error(json_object["error"])
        else:
            self.error = ""
    
    def get_data(self):
        '''
        @return JsonObject the data part of the API response
        '''
        return self.data
    
    
    def set_data(self, data):
        '''
        @param $data the data to include in the API response
        '''
        self.data = json.loads(data)
    

    def get_error(self):
        '''
        @return Error the error part of the API response, consisting of an error code and an error message
        '''
        return self.error

    
    def set_error(self, error):
        '''
        @param $error an error to include in the API response
        '''
        self.error = Error(error)


    def to_json(self):
        '''
        @return a Json object with the data and error parts set if they exist
        '''
        json_response = {}
        
        if hasattr(self, "data"):
            json_response["data"] = self.data
        
        if hasattr(self, "error"):
            json_response["error"] = self.error

        return json_response;  


class Latch(object):
    
    API_HOST = "latch.elevenpaths.com";
    API_PORT = 443;
    API_HTTPS = True
    API_PROXY = None;
    API_PROXY_PORT = None;
    API_CHECK_STATUS_URL = "/api/0.6/status";
    API_PAIR_URL = "/api/0.6/pair";
    API_PAIR_WITH_ID_URL = "/api/0.6/pairWithId";
    API_UNPAIR_URL = "/api/0.6/unpair";

    
    AUTHORIZATION_HEADER_NAME = "Authorization";
    DATE_HEADER_NAME = "X-11Paths-Date";
    AUTHORIZATION_METHOD = "11PATHS";
    AUTHORIZATION_HEADER_FIELD_SEPARATOR = " ";

    UTC_STRING_FORMAT = "%Y-%m-%d %H:%M:%S";
    
    X_11PATHS_HEADER_PREFIX = "X-11paths-";
    X_11PATHS_HEADER_SEPARATOR = ":";
    
    @staticmethod
    def set_host(host):
        '''
        @param $host The host to be connected with (http://hostname) or (https://hostname)
        '''
        if host.startswith("http://"):
            Latch.API_HOST = host[len("http://"):]
            Latch.API_PORT = 80
            Latch.API_HTTPS = False
        elif host.startswith("https://"):
            Latch.API_HOST = host[len("https://"):]
            Latch.API_PORT = 443
            Latch.API_HTTPS = True

    @staticmethod
    def set_proxy(proxy, port):
        '''
        Enable using a Proxy to connect through
        @param $proxy The proxy server
        @param $port The proxy port number
        '''
        Latch.API_PROXY = proxy
        Latch.API_PROXY_PORT = port

    @staticmethod
    def get_part_from_header(part, header):
        '''
        The custom header consists of three parts, the method, the appId and the signature.
        This method returns the specified part if it exists.
        @param $part The zero indexed part to be returned
        @param $header The HTTP header value from which to extract the part
        @return string the specified part from the header or an empty string if not existent
        '''
        if (header):
            parts = header.split(Latch.AUTHORIZATION_HEADER_FIELD_SEPARATOR);
            if(len(parts) >= part):
                return parts[part]
        return ""
    
    @staticmethod
    def get_auth_method_from_header(authorizationHeader):
        '''
        @param $authorizationHeader Authorization HTTP Header
        @return string the Authorization method. Typical values are "Basic", "Digest" or "11PATHS"
        '''
        return Latch.get_part_from_header(0, authorizationHeader)

    @staticmethod  
    def get_app_id_from_header(authorizationHeader):
        '''
        @param $authorizationHeader Authorization HTTP Header
        @return string the requesting application Id. Identifies the application using the API
        '''
        return Latch.get_part_from_header(1, authorizationHeader)

    @staticmethod
    def get_signature_from_header(authorizationHeader):
        '''
        @param $authorizationHeader Authorization HTTP Header
        @return string the signature of the current request. Verifies the identity of the application using the API
        '''
        return Latch.get_part_from_header(2, authorizationHeader)


    def __init__(self, appId, secretKey):
        '''
        Create an instance of the class with the Application ID and secret obtained from Eleven Paths
        @param $appId
        @param $secretKey
        '''
        self.appId = appId
        self.secretKey = secretKey




    def _http_get(self, url, xHeaders=None):
        '''
        HTTP GET Request to the specified API endpoint
        @param $string $url
        @param $string $xHeaders
        @return LatchResponse
        '''
        try:
            # Try to use the new Python3 HTTP library if available
            import http.client as http
        except:
            # Must be using Python2 so use the appropriate library
            import httplib as http

        authHeaders = self.authentication_headers("GET", url, xHeaders)
        #print(headers)
        if Latch.API_PROXY != None:
            if Latch.API_HTTPS:
                conn = http.HTTPSConnection(Latch.API_PROXY, Latch.API_PROXY_PORT)
                conn.set_tunnel(Latch.API_HOST, Latch.API_PORT) 
            else: 
                conn = http.HTTPConnection(Latch.API_PROXY, Latch.API_PROXY_PORT)
                url = "http://" + Latch.API_HOST + url                  
        else:
            if Latch.API_HTTPS:
                conn = http.HTTPSConnection(Latch.API_HOST, Latch.API_PORT)
            else: 
                conn = http.HTTPConnection(Latch.API_HOST, Latch.API_PORT)

        try:
            conn.request("GET", url, headers=authHeaders)
            response = conn.getresponse()
           
            responseData = response.read().decode('utf8')
            #print("response:" + responseData)
            conn.close();
            ret = LatchResponse(responseData)
        except:
            ret = LatchResponse("{}")

        return ret
        

    def pairWithId(self, accountId):
        return self._http_get(self.API_PAIR_WITH_ID_URL + "/" + accountId)

    def pair(self, token):
        return self._http_get(self.API_PAIR_URL + "/" + token)

    def status(self, accountId):
        return self._http_get(self.API_CHECK_STATUS_URL + "/" + accountId)

    def operationStatus(self, accountId, operationId):
        return self._http_get(self.API_CHECK_STATUS_URL + "/" + accountId + "/op/" + operationId)

    def unpair(self, accountId):
        return self._http_get(self.API_UNPAIR_URL + "/" + accountId)



    
    def sign_data(self, data):
        '''
        @param $data the string to sign
        @return string base64 encoding of the HMAC-SHA1 hash of the data parameter using {@code secretKey} as cipher key.
        '''
        from hashlib import sha1
        import hmac
        import binascii
        
        sha1Hash = hmac.new(self.secretKey.encode(), data.encode(), sha1)
        return binascii.b2a_base64(sha1Hash.digest())[:-1].decode('utf8')
       
    
    def authentication_headers(self, HTTPMethod, queryString, xHeaders=None, utc=None):
        '''
        Calculate the authentication headers to be sent with a request to the API
        @param $HTTPMethod the HTTP Method, currently only GET is supported
        @param $queryString the urlencoded string including the path (from the first forward slash) and the parameters
        @param $xHeaders HTTP headers specific to the 11-paths API. null if not needed.
        @param $utc the Universal Coordinated Time for the Date HTTP header
        @return array a map with the Authorization and Date headers needed to sign a Latch API request
        '''
        if (not utc):
            utc = Latch.get_current_UTC()
        
        utc = utc.strip()
        
        #logging.debug(HTTPMethod);
        #logging.debug(queryString);
        #logging.debug(utc);

        stringToSign = (HTTPMethod.upper().strip() + "\n" + 
                        utc + "\n" + 
                        self.get_serialized_headers(xHeaders) + "\n" +
                        queryString.strip())
               
        authorizationHeader = (Latch.AUTHORIZATION_METHOD + Latch.AUTHORIZATION_HEADER_FIELD_SEPARATOR + 
                               self.appId + Latch.AUTHORIZATION_HEADER_FIELD_SEPARATOR + 
                               self.sign_data(stringToSign))

        headers = dict()
        headers[Latch.AUTHORIZATION_HEADER_NAME] = authorizationHeader;
        headers[Latch.DATE_HEADER_NAME] = utc;
        return headers

    
    def get_serialized_headers(self, xHeaders):
        '''
        Prepares and returns a string ready to be signed from the 11-paths specific HTTP headers received
        @param $xHeaders a non neccesarily ordered map (array without duplicates) of the HTTP headers to be ordered.
        @return string The serialized headers, an empty string if no headers are passed, or None if there's a problem such as non 11paths specific headers
        '''
        if (xHeaders):
            headers = dict((k.lower(), v) for k, v in xHeaders.iteritems())
            headers.sort()
            serializedHeaders = ""
            for key, value in headers:
                if (not key.startsWith(Latch.X_11PATHS_HEADER_PREFIX.lower())):
                    logging.error("Error serializing headers. Only specific " + Latch.X_11PATHS_HEADER_PREFIX + " headers need to be singed")
                    return None
                serializedHeaders += key + Latch.X_11PATHS_HEADER_SEPARATOR + value + " "
            return serializedHeaders.strip()
        else:
            return ""
    
    @staticmethod
    def get_current_UTC():
        '''
        @return a string representation of the current time in UTC to be used in a Date HTTP Header
        '''
        import time
        return time.strftime(Latch.UTC_STRING_FORMAT, time.gmtime())
    

File: lib/latch.c
/*
 * This library implements all functions of Latch API.
 * Copyright (C) 2013 Eleven Paths

 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2.1 of the License, or (at your option) any later version.
 
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */

#include <ctype.h>
#include <stdarg.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

#include <curl/curl.h>
#include <curl/easy.h>

#include <openssl/hmac.h>
#include <openssl/bio.h>
#include <openssl/evp.h>
#include <openssl/buffer.h>

#define AUTHORIZATION_HEADER_NAME "Authorization"
#define DATE_HEADER_NAME "X-11Paths-Date"
#define AUTHORIZATION_METHOD "11PATHS"
#define AUTHORIZATION_HEADER_FIELD_SEPARATOR " "
#define UTC_STRING_FORMAT "%Y-%m-%d %H:%M:%S"

#define API_CHECK_STATUS_URL "/api/0.9/status"
#define API_PAIR_URL "/api/0.9/pair"
#define API_PAIR_WITH_ID_URL "/api/0.9/pairWithId"
#define API_UNPAIR_URL "/api/0.9/unpair"
#define API_LOCK_URL "/api/0.9/lock"
#define API_UNLOCK_URL "/api/0.9/unlock"
#define API_HISTORY_URL "/api/0.9/history"
#define API_OPERATION_URL "/api/0.9/operation"

#define HTTP_METHOD_GET "GET"
#define HTTP_METHOD_POST "POST"
#define HTTP_METHOD_PUT "PUT"
#define HTTP_METHOD_DELETE "DELETE"

#define HTTP_PARAM_LOCK_ON_REQUEST "lock_on_request"
#define HTTP_PARAM_NAME "name"
#define HTTP_PARAM_PARENTID "parentId"
#define HTTP_PARAM_TWO_FACTOR "two_factor"

typedef struct curl_response_buffer {
    char *buffer;
    size_t size;
} curl_response_buffer;

typedef struct http_param {
    char *name;
    char *value;
} http_param;

/*
 * Function to handle stuff from HTTP response.
 * 
 * @param buf- Raw buffer from libcurl.
 * @param len- number of indexes
 * @param size- size of each index
 * @param userdata- any extra user data needed
 * @return Number of bytes actually handled. If different from len * size, curl will throw an error
 */
static int writeFn(void* buf, size_t len, size_t size, void* userdata) {

    size_t realsize = len * size;
    curl_response_buffer *response = (curl_response_buffer*)userdata;

    response->buffer = realloc(response->buffer, response->size + realsize + 1);

    memcpy(&(response->buffer[response->size]), buf, realsize);
    response->size += realsize;
    response->buffer[response->size] = '\0';

    return realsize;

}

/*
 * Function to encode a string in Base64 format
 * 
 * @param input- string to encode
 * @param length- string length
 * @return encoded string in Base64 format
 */

char* base64encode(const unsigned char *input, int length) {
	BIO *bmem, *b64;
	BUF_MEM *bptr;

	b64 = BIO_new(BIO_f_base64());
	bmem = BIO_new(BIO_s_mem());
	b64 = BIO_push(b64, bmem);
	BIO_write(b64, input, length);
	BIO_flush(b64);
	BIO_get_mem_ptr(b64, &bptr);

	char *buff = (char *)malloc(bptr->length);
	memcpy(buff, bptr->data, bptr->length - 1);
	buff[bptr->length - 1] = 0;

	BIO_free_all(b64);

	return buff;
}

char toHex(char code) {
    static char hex[] = "0123456789ABCDEF";
    return hex[code & 15];
}

/*
 * Function to percent-encode a string
 *
 * Based on http://www.geekhideout.com/downloads/urlcode.c
 */
char* urlEncode(const char* str, int space2Plus) {

    char* buf = NULL;
    char* pbuf = NULL;
    const char* pstr = str;

    if ((str != NULL) && ((buf = malloc(strlen(str) * 3 + 1)) != NULL)) {
        pbuf = buf;
        while (*pstr) {
            if (isalnum(*pstr) || *pstr == '-' || *pstr == '_' || *pstr == '.' || *pstr == '~') {
                *pbuf++ = *pstr;
            } else if (*pstr == ' ' && space2Plus) {
                *pbuf++ = '+';
            } else {
                *pbuf++ = '%';
                *pbuf++ = toHex(*pstr >> 4);
                *pbuf++ = toHex(*pstr & 15);
            }
            pstr++;
        }
        *pbuf = '\0';
    }

    return buf;

}

/*
 * Function to calculate the HMAC hash (SHA1) of a string. Returns a Base64 value of the hash
 * 
 * @param pSecretKey- secret key
 * @param pData- original data to calculate the HMAC
 * @return HMAC in Base64 format
 */
char* sign_data(const char* pSecretKey, const char* pData) {
    unsigned char* digest;

    digest = HMAC(EVP_sha1(), pSecretKey, strlen(pSecretKey), (unsigned char*) pData, strlen(pData), NULL, NULL);
    return base64encode(digest, 20);
}

int nosignal = 0;
int timeout = 2;
const char* AppId;
const char* SecretKey;
const char* Host = "https://latch.elevenpaths.com";
const char* Proxy;
const char* tlsCAFile = NULL;
const char* tlsCAPath = NULL;
const char* tlsCRLFile = NULL;

void init(const char* pAppId, const char* pSecretKey) {
	AppId = pAppId;
	SecretKey = pSecretKey;
}

void setHost(const char* pHost){
	Host = pHost;
}

/*
 * Enable using a Proxy to connect to Latch Server
 */
void setProxy(const char* pProxy){
	Proxy = pProxy;
}

void setTimeout(const int iTimeout)
{
    timeout = iTimeout;
}

/*
 * If called with iNoSignal 1, CURLOPT_NOSIGNAL will be set to 1
 */
void setNoSignal(const int iNoSignal)
{
    nosignal = iNoSignal;
}

void setTLSCAFile(const char* pTLSCAFile)
{
    tlsCAFile = pTLSCAFile;
}

void setTLSCAPath(const char* pTLSCAPath)
{
    tlsCAPath = pTLSCAPath;
}

void setTLSCRLFile(const char* pTLSCRLFile)
{
    tlsCRLFile = pTLSCRLFile;
}

void authenticationHeaders(const char* pHTTPMethod, const char* pQueryString, char* pHeaders[], const char *pBody) {

	char* authHeader, *dateHeader, *stringToSign, *b64hash;
	char utc[20];
	time_t timer;
	struct tm tm_info;
	int len = 0;

	time(&timer);
	gmtime_r(&timer, &tm_info);
	strftime(utc, 20, UTC_STRING_FORMAT, &tm_info);

	if (pBody == NULL) {
	    len = strlen(pHTTPMethod) + strlen(utc) + strlen(pQueryString) + 4;
	} else {
	    len = strlen(pHTTPMethod) + strlen(utc) + strlen(pQueryString) + strlen(pBody) + 5;
	}

	stringToSign = malloc(len);

	if (pBody == NULL) {
	    snprintf(stringToSign, len, "%s\n%s\n\n%s", pHTTPMethod, utc, pQueryString);
	} else {
	    snprintf(stringToSign, len, "%s\n%s\n\n%s\n%s", pHTTPMethod, utc, pQueryString, pBody);
	}

	b64hash = sign_data(SecretKey, stringToSign);

	len = strlen(AUTHORIZATION_HEADER_NAME) + strlen(AUTHORIZATION_METHOD) + strlen(AppId) + strlen(b64hash) + 5;
	authHeader = malloc(len);
	snprintf(authHeader, len, "%s: %s %s %s", AUTHORIZATION_HEADER_NAME, AUTHORIZATION_METHOD, AppId, b64hash);

	len = strlen(DATE_HEADER_NAME) + strlen(utc) + 3;
	dateHeader = malloc(len);
	snprintf(dateHeader, len, "%s: %s", DATE_HEADER_NAME, utc);

	pHeaders[0] = authHeader;
	pHeaders[1] = dateHeader;

    free(stringToSign);
    free(b64hash);

}

/*
 * Perform a GET request to the specified URL of the Latch API
 * @param pUrl- requested URL including host
 */
char* http_proxy(const char* pMethod, const char* pUrl, const char* pBody) {

	char* headers[2];
	curl_response_buffer response;
	char error_message[CURL_ERROR_SIZE];
	CURL* pCurl = curl_easy_init();
	int res = -1;
	int i = 0;
	struct curl_slist* chunk = NULL;
	char* hostAndUrl;

	if (!pCurl) {
		return NULL;
	}

    response.buffer = malloc(1*sizeof(char));
    response.size = 0;
    response.buffer[response.size] = '\0';

	authenticationHeaders(pMethod, pUrl, headers, pBody);
	for (i=0; i<(sizeof(headers)/sizeof(char*)); i++) {
		chunk = curl_slist_append(chunk, headers[i]);
	}

    free(headers[0]);
    free(headers[1]);

	hostAndUrl = malloc(strlen(Host) + strlen(pUrl) + 1);
	strcpy(hostAndUrl, Host);
	strcat(hostAndUrl, pUrl);

	curl_easy_setopt(pCurl, CURLOPT_URL, hostAndUrl);
	curl_easy_setopt(pCurl, CURLOPT_HTTPHEADER, chunk);
	curl_easy_setopt(pCurl, CURLOPT_WRITEFUNCTION, writeFn);
	curl_easy_setopt(pCurl, CURLOPT_WRITEDATA, &response);
	curl_easy_setopt(pCurl, CURLOPT_NOPROGRESS, 1); // we don't care about progress
	curl_easy_setopt(pCurl, CURLOPT_FAILONERROR, 1);

	curl_easy_setopt(pCurl, CURLOPT_CUSTOMREQUEST, pMethod);

	if ((strncmp(pMethod, HTTP_METHOD_POST, strlen(HTTP_METHOD_POST)) == 0) || (strncmp(pMethod, HTTP_METHOD_PUT, strlen(HTTP_METHOD_PUT)) == 0)) {
	    curl_easy_setopt(pCurl, CURLOPT_POSTFIELDS, pBody);
	    if (pBody == NULL) {
	        curl_easy_setopt(pCurl, CURLOPT_POSTFIELDSIZE, 0);
	    }
	}

	if(Proxy != NULL){
		curl_easy_setopt(pCurl, CURLOPT_PROXY, Proxy);
	}

	// we don't want to leave our user waiting at the login prompt forever
	curl_easy_setopt(pCurl, CURLOPT_TIMEOUT, timeout);

	// SSL needs 16k of random stuff. We'll give it some space in RAM.
	curl_easy_setopt(pCurl, CURLOPT_RANDOM_FILE, "/dev/urandom");
	curl_easy_setopt(pCurl, CURLOPT_SSL_VERIFYPEER, 1);
	curl_easy_setopt(pCurl, CURLOPT_SSL_VERIFYHOST, 2);
	curl_easy_setopt(pCurl, CURLOPT_USE_SSL, CURLUSESSL_ALL);

	// error message when curl_easy_perform return non-zero
	curl_easy_setopt(pCurl, CURLOPT_ERRORBUFFER, error_message);

	// Optional if setNoSignal(1)
	// Avoid crashing if multithread, DNS timeout and libcurl < 7.32.0
	// Blocks with standard resolver (doesn't apply the timeout)

	if (nosignal == 1) {
	    curl_easy_setopt(pCurl, CURLOPT_NOSIGNAL, 1);
	}

	if (tlsCAFile != NULL) {
	    curl_easy_setopt(pCurl, CURLOPT_CAINFO, tlsCAFile);
	    curl_easy_setopt(pCurl, CURLOPT_CAPATH, NULL);
	}
	else {
	    if (tlsCAPath != NULL) {
	        curl_easy_setopt(pCurl, CURLOPT_CAINFO, NULL);
	        curl_easy_setopt(pCurl, CURLOPT_CAPATH, tlsCAPath);
	    }
	}

	if (tlsCRLFile != NULL) {
	    curl_easy_setopt(pCurl, CURLOPT_CRLFILE, tlsCRLFile);
	}

	// synchronous, but we don't really care
	res = curl_easy_perform(pCurl);

	curl_easy_cleanup(pCurl);
    curl_slist_free_all(chunk);
    free(hostAndUrl);

    if (res != CURLE_OK) {
        free(response.buffer);
        return NULL;
    }
    else {
        return response.buffer;
    }

}

char* build_string(int argc, char** argv) {

    int i = 0;
    int len = 1;
    char *rv = NULL;

    for (i = 0; i < argc; i++) {
        if (argv[i] != NULL) {
            len += strlen(argv[i]);
        }
    }

    if ((rv = (char *) malloc(len)) != NULL) {
        *rv = '\0';
        for (i = 0; i < argc; i++) {
            if (argv[i] != NULL) {
                strncat(rv, argv[i], strlen(argv[i]));
            }
        }
    }

    return rv;

}

char* build_url(int argc, ...) {

    int i = 0;
    char *tokens[2 * argc - 1];
    char *response = NULL;
    va_list args;

    va_start(args, argc);
    for (i = 0; i < 2 * argc - 1; i = i + 2) {
        if (i == 0) {
            tokens[i] = va_arg(args, char*);
        } else {
            tokens[i] = urlEncode(va_arg(args, char*), 0);
        }
        if (i < 2 * argc - 1) {
            if (tokens[i] == NULL) {
                tokens[i + 1] = NULL;
            } else {
                tokens[i + 1] = "/";
            }
        }
    }
    va_end(args);

    response = build_string(2 * argc - 1, tokens);

    for (i = 0; i < 2 * argc - 1; i = i + 2) {
        if (i != 0) {
            free(tokens[i]);
        }
    }

    return response;

}

char* build_url_v(int argc, va_list args) {

    int i = 0;
    char *tokens[2 * argc - 1];
    char *response = NULL;

    for (i = 0; i < 2 * argc - 1; i = i + 2) {
        if (i == 0) {
            tokens[i] = va_arg(args, char*);
        } else {
            tokens[i] = urlEncode(va_arg(args, char*), 0);
        }
        if (i < 2 * argc - 1) {
            if (tokens[i] == NULL) {
                tokens[i + 1] = NULL;
            } else {
                tokens[i + 1] = "/";
            }
        }
    }

    response = build_string(2 * argc - 1, tokens);

    for (i = 0; i < 2 * argc - 1; i = i + 2) {
        if (i != 0) {
            free(tokens[i]);
        }
    }

    return response;

}

char* build_querystring(int argc, http_param* params) {

    int i = 0;
    int j = 0;
    char *rv = NULL;
    char *tokens[4 * argc];

    for (i = 0, j = 0; i < 4 * argc; i = i + 4, j++) {
        if (params[j].name != NULL && params[j].value != NULL) {
            tokens[i] = params[j].name;
            tokens[i + 1] = "=";
            tokens[i + 2] = params[j].value;
            tokens[i + 3] = "&";
        } else {
            tokens[i] = NULL;
            tokens[i + 1] = NULL;
            tokens[i + 2] = NULL;
            tokens[i + 3] = NULL;
        }
    }

    rv = build_string(4 * argc, tokens);
    rv[strlen(rv) - 1] = '\0';

    return rv;

}

char* operation(const char* pMethod, int nParams, http_param* params, int nUrlTokens, ...) {

    int i = 0;
    int valid = 1;
    char *response = NULL;
    char *url = NULL;
    char *body = NULL;
    va_list args;

    va_start(args, nUrlTokens);
    for (i = 0; i < nUrlTokens; i++) {
        if (va_arg(args, char*) == NULL) {
            valid = 0;
        }
    }
    va_end(args);

    va_start(args, nUrlTokens);

    if (valid && ((url = build_url_v(nUrlTokens, args)) != NULL)) {

        if (nParams > 0) {
            body = build_querystring(nParams, params);
        }

        response = http_proxy(pMethod, url, body);

        free(url);
        free(body);

    }

    va_end(args);

    return response;

}

char* pairWithId(const char* pAccountId) {
    return operation(HTTP_METHOD_GET, 0, NULL, 2, API_PAIR_WITH_ID_URL, pAccountId);
}

char* pair(const char* pToken) {
    return operation(HTTP_METHOD_GET, 0, NULL, 2, API_PAIR_URL, pToken);
}

char* status(const char* pAccountId) {
    return operation(HTTP_METHOD_GET, 0, NULL, 2, API_CHECK_STATUS_URL, pAccountId);
}

char* operationStatus(const char* pAccountId, const char* pOperationId) {
    return operation(HTTP_METHOD_GET, 0, NULL, 4, API_CHECK_STATUS_URL, pAccountId, "op", pOperationId);
}

char* unpair(const char* pAccountId) {
    return operation(HTTP_METHOD_GET, 0, NULL, 2, API_UNPAIR_URL, pAccountId);
}

char* lock(const char* pAccountId) {
    return operation(HTTP_METHOD_POST, 0, NULL, 2, API_LOCK_URL, pAccountId);
}

char* operationLock(const char* pAccountId, const char* pOperationId) {
    return operation(HTTP_METHOD_POST, 0, NULL, 4, API_LOCK_URL, pAccountId, "op", pOperationId);
}

char* unlock(const char* pAccountId) {
    return operation(HTTP_METHOD_POST, 0, NULL, 2, API_UNLOCK_URL, pAccountId);
}

char* operationUnlock(const char* pAccountId, const char* pOperationId) {
    return operation(HTTP_METHOD_POST, 0, NULL, 4, API_UNLOCK_URL, pAccountId, "op", pOperationId);
}

char* history(const char* pAccountId) {
    return operation(HTTP_METHOD_GET, 0, NULL, 2, API_HISTORY_URL, pAccountId);
}

char* timePeriodHistory(const char* pAccountId, time_t from, time_t to) {

    char sFrom[14];
    char sTo[14];

    if (from == 0) {
        snprintf(sFrom, 14, "%lld", (long long)from);
    } else {
        snprintf(sFrom, 14, "%lld000", (long long)from);
    }

    if (to == 0) {
        snprintf(sTo, 14, "%lld", (long long)to);
    } else {
        snprintf(sTo, 14, "%lld000", (long long)to);
    }

    return operation(HTTP_METHOD_GET, 0, NULL, 4, API_HISTORY_URL, pAccountId, sFrom, sTo);

}

char* operationCreate(const char* pParentId, const char* pName, const char* pTwoFactor, const char* pLockOnRequest) {

    char *response = NULL;
    http_param params[4];

    if (pParentId != NULL && pName != NULL) {

        params[0].name = HTTP_PARAM_LOCK_ON_REQUEST;
        params[0].value = pLockOnRequest == NULL ? NULL : urlEncode(pLockOnRequest, 1);
        params[1].name = HTTP_PARAM_NAME;
        params[1].value = urlEncode(pName, 1);
        params[2].name = HTTP_PARAM_PARENTID;
        params[2].value = urlEncode(pParentId, 1);
        params[3].name = HTTP_PARAM_TWO_FACTOR;
        params[3].value = pTwoFactor == NULL ? NULL : urlEncode(pTwoFactor, 1);

        response = operation(HTTP_METHOD_PUT, 4, params, 1, API_OPERATION_URL);

        free(params[0].value);
        free(params[1].value);
        free(params[2].value);
        free(params[3].value);

    }

    return response;

}

char* operationUpdate(const char* pOperationId, const char* pName, const char* pTwoFactor, const char* pLockOnRequest) {

    char *response = NULL;
    http_param params[3];

    if (pOperationId != NULL && (pName != NULL || pTwoFactor != NULL || pLockOnRequest != NULL)) {

        params[0].name = HTTP_PARAM_LOCK_ON_REQUEST;
        params[0].value = pLockOnRequest == NULL ? NULL : urlEncode(pLockOnRequest, 1);
        params[1].name = HTTP_PARAM_NAME;
        params[1].value = pName == NULL ? NULL : urlEncode(pName, 1);
        params[2].name = HTTP_PARAM_TWO_FACTOR;
        params[2].value = pTwoFactor == NULL ? NULL : urlEncode(pTwoFactor, 1);

        response = operation(HTTP_METHOD_POST, 3, params, 2, API_OPERATION_URL, pOperationId);

        free(params[0].value);
        free(params[1].value);
        free(params[2].value);

    }

    return response;

}

char* operationRemove(const char* pOperationId) {
    return operation(HTTP_METHOD_DELETE, 0, NULL, 2, API_OPERATION_URL, pOperationId);
}

char* operationGet(const char* pOperationId) {
    return operation(HTTP_METHOD_GET, 0, NULL, 2, API_OPERATION_URL, pOperationId);
}

char* operationsGet() {
    return operation(HTTP_METHOD_GET, 0, NULL, 1, API_OPERATION_URL);
}
